### PR TITLE
feat(google-access): testing usersync and Google Data Access [PXD-2280]

### PR DESCRIPTION
**Pull request by Avantol13**
testing usersync and Google Data Access access to google bucket after usersync

* can run usersync job from tests
* actually hits Google's API to try and read a file from a bucket for different users

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
Jenkins Build 16 : time taken 24 min

**Comment by PlanXCyborg** [comment]
Jenkins Build 17 : time taken 18 min

**Comment by PlanXCyborg** [comment]
Jenkins Build 18 : time taken 15 min

**Comment by PlanXCyborg** [comment]
Jenkins Build 19 : time taken 39 min

File: suites/apis/googleDataAccessTest.js [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+  const pathToCreds0KeyFile = creds0Key + '.json';
+  const pathToCreds1KeyFile = creds1Key + '.json';
+  const pathToCreds2KeyFile = creds2Key + '.json';
+
+  fs.writeFile(pathToCreds0KeyFile, JSON.stringify(tempCreds0Res.body), function(err) {
+    if (err) {
+      return console.log(err);
+    }
+    console.log(`Google creds file ${pathToCreds0KeyFile} saved!`);
+  });
```
Comment (paulineribeyre) [inline suites/apis/googleDataAccessTest.js]: there is a util function that makes sure to wait for the file to be written before returning:
`await files.createTmpFile(pathToCreds0KeyFile, JSON.stringify(tempCreds0Res.body));`

File: suites/apis/googleDataAccessTest.js [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+  const deleteServiceAccount1Res = await google.deleteServiceAccount(
+    tempCreds1Res.body.project_id, tempCreds1Res.body.client_email);
+  const deleteServiceAccount2Res = await google.deleteServiceAccount(
+    tempCreds2Res.body.project_id, tempCreds2Res.body.client_email);
+
+  console.log('deleting temporary google credentials file');
+  fs.unlink(pathToCreds0KeyFile, function(err) {
+    if (err) throw err;
+    console.log(`${pathToCreds0KeyFile} deleted!`);
+  });
```
Comment (paulineribeyre) [inline suites/apis/googleDataAccessTest.js]: you can also use the util function here `files.deleteFile(pathToCreds0KeyFile);`
we could add a log for success to both util functions

File: suites/apis/googleDataAccessTest.js (hunk 1 above)
Comment (Avantol13) [reply suites/apis/googleDataAccessTest.js]: oh, nice. thanks!

File: utils/commons.js [hunk 3]
```diff
@@ -27,6 +30,14 @@ class Commons{
     }
   }
 
+  // these files exist in cloud automation repo, check there for details
+  static get userAccessFiles() {
+    return {
+      newUserAccessFile1: 'test1_user.yaml',
```
Comment (paulineribeyre) [inline utils/commons.js]: file1 is never used
Also we could change the logic for this - from our previous discussion:
"we need a function that modifies the backed up YAML with whats in the test2 file (changed access for test users) and then we can run useryaml job with that file"
(see `setUserYaml` function)
Maybe a refactoring idea for later~~

Comment (paulineribeyre) [reply utils/commons.js]: `backupUserYaml()` is not used for now but let's keep it in case we want to improve this

File: utils/google.js [hunk 4]
```diff
@@ -90,6 +92,32 @@ const googleApp = {
  * Exported google util functions
  */
 module.exports = {
+  async getFileFromBucket(googleProject, pathToCredsKeyFile, bucketName, fileName) {
```
Comment (paulineribeyre) [inline utils/google.js]: not good error handling - the fixed function is in my branch https://github.com/uc-cdis/gen3-qa/blob/feat/google-expire/utils/google.js#L158
we can copy it here or just wait for my PR to be merged

File: suites/apis/googleDataAccessTest.js [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+
+  chai.expect(User2signedUrlTest2Res,
+    '2nd sync: Check that User2 could NOT get a signed URL to read file in test. FAILED.'
+  ).to.have.property('statusCode', 401);
+
+  // SECOND RUN
+  //  - Check signed URLs from FIRST RUN
+  chai.expect(User0AccessRemovedQA,
+    "Make sure signed URL from User0 CANNOT access QA data again. FAILED."
+  ).to.not.equal(fence.props.googleBucketInfo.QA.fileContents);
```
Comment (paulineribeyre) [inline suites/apis/googleDataAccessTest.js]: i think `equal('some error msg')` would be safer than `not.equal('expected contents')`

File: services/apis/fence/fenceSequences.js [hunk 6]
```diff
@@ -50,6 +51,20 @@ module.exports = {
...
+   * wan't a linked account to begin with
+   * @param {User} userAcct - user to delete link for
+   * @returns {Promise<void>}
+   */
+  async forceUnlinkGoogleAcct(userAcct) {
+    const unlinkRes = await fenceTasks.unlinkGoogleAcct(userAcct);
+    expect(unlinkRes,
+      'response from unlinking Google Account does not have expected statusCode property'
+    ).to.have.property('statusCode');
+    // can be 200 or 404
```
[the remark this answers is not in the document]
Comment (Avantol13) [reply services/apis/fence/fenceSequences.js]: https://www.chaijs.com/api/bdd/#method_oneof

File: services/apis/fence/fenceTasks.js [hunk 7]
```diff
@@ -181,15 +197,14 @@ module.exports = {
 
   /**
    * Goes through the full, proper process for linking a google account assuming env
-   * is set to mock Google response
+   * is set to mock Google responsenew Error
```
Comment (paulineribeyre) [inline services/apis/fence/fenceTasks.js]: what's that :p

File: suites/apis/googleDataAccessTest.js [hunk 8]
```diff
@@ -471,11 +471,11 @@ Scenario('Test Google Data Access (signed urls and temp creds) @reqGoogle @googl
   //  - Check signed URLs from FIRST RUN
   chai.expect(User0AccessRemovedQA,
     "Make sure signed URL from User0 CANNOT access QA data again. FAILED."
-  ).to.not.equal(fence.props.googleBucketInfo.QA.fileContents);
+  ).to.contain('AccessDenied');
 
   chai.expect(User1AccessRemovedQA,
     "Make sure signed URL from User1 CANNOT access QA data again. FAILED."
-  ).to.not.equal(fence.props.googleBucketInfo.QA.fileContents);
+  ).to.contain('AccessDenied');
```
Comment (paulineribeyre) [inline suites/apis/googleDataAccessTest.js]: where does `AccessDenied` come from? shouldn't it return `GOOGLE_FILE_FROM_URL_ERROR`?

File: services/apis/fence/fenceTasks.js [hunk 9]
```diff
@@ -71,6 +73,8 @@ async function loginGoogle(googleCreds) {
  * fence Tasks
  */
 module.exports = {
+  GOOGLE_FILE_FROM_URL_ERROR,
```
Comment (paulineribeyre) [inline services/apis/fence/fenceTasks.js]: it's not used outside of this file so no need to export it

File: services/apis/fence/fenceTasks.js (hunk 7 above)
Comment (Avantol13) [reply services/apis/fence/fenceTasks.js]: whoops

File: suites/apis/googleDataAccessTest.js (hunk 8 above)
Comment (Avantol13) [reply suites/apis/googleDataAccessTest.js]: So turns out google will still allow accessing the signed url but they send back `AccessDenied` in the response

**Pull request merged**